### ISSUE TITLE
GroupBy using grouper function that returns complex type doesn not work

### ISSUE THREAD

**Issue by Stijnlb** [kutyel/linq.ts#68]
When performing a GroupBy on a List<<T>> using a grouper function that returns a "complex" anonymous type, GroupBy returns a flat array of objects (untyped "T" objects) instead of a list of key/value pairs as it does when grouping with a grouper function that returns a primitive type.

e.g with a type.

```typescript
interface IReferral {
    origin: string;
    modalityType: string;
    qtyReferrals: number;
}
```

Not working:

```typescript
let referralsByOrigin = new List<IReferral>(data).GroupBy(key => ({ org: key.origin, modType: key.modalityType}), element => element);
```
I am currently working around this by creating a primitive returning grouper  that returns a concatenation of my grouping elements like below, and then, when processing the result, retrieving the actual grouping elements from the first item. (instead of retrieving them as "menbers" of the grouper result (key) :

```typescript
let groupedData = new List<IReferral>(result).GroupBy(key => key.origin + '_' + key.modalityType , element => element);

for (var key in groupedData) {
  // count the total nr of referrals
  var totalReferrals = groupedData[key].reduce((a, b) => a + b.qtyReferrals, 0);
  var firstItem : IReferral = new List<IReferral> (groupedData[key]).First();
  this.chartData.push([firstItem.origin, firstItem.modalityType, totalReferrals]);
}
```
a fix to this would be great!

**Comment by kutyel** [comment]
@Stijnlb Is this the same bug as in #56?

**Comment by Stijnlb** [comment]
@kutyel  No, to my understanding it is not ... #56 is about the person not liking the return type and rather getting IGrouping objects returned ... In my example, GroupBy is simply not working when using a grouper function that returns an anonymous type .... it just returns a flat untyped array corresponding to the original List of T ...

When using a grouper function that returns a primitive type (string, number, ...) as grouping key it does work ... (that's why I ended up concatenating my grouping terms in a primitive string instead of using an anonymous type with multiple members ...)

**Comment by kutyel** [comment]
OK, will work on this when I have time. Can you provide a more detail example with source data and maybe a **.NET Fiddle** that shows how this works on .NET? Thanks!

**Comment by abbasmhd** [comment]
@Stijnlb can you provide a test case or an example please

**Comment by marlon-tucker** [comment]
This will never work, the object being returned is "new" and therefore always unique. In C# I believe this works due to anonymous objects implementing a simplistic IEquality interface which hashes the properties. This is not achievable in javascript.

You *could* inspect the mapped return type to see if it contains a "equals" function, and use the return value from that function instead, but in that case, the mapping function is better placed to call it instead.

In short, this is not achievable with the current set of tools javascript provides.

**Comment by kutyel** [comment]
I'll close this for now since as discussed in the thread I don't think this is possible yet due to the nature of JavaScript itself 🤷‍♂️
